src/storage/mutate/AddEdgesProcessor.cpp

```diff
@@ -13,27 +13,133 @@ namespace nebula {
-                    << ", EdgeType: " << edge.key.edge_type << ", EdgeRanking: " << edge.key.ranking
-                    << ", VertexID: " << edge.key.dst << ", EdgeVersion: " << version;
-            auto key = NebulaKeyUtils::edgeKey(partId, edge.key.src, edge.key.edge_type,
-                                               edge.key.ranking, edge.key.dst, version);
-            data.emplace_back(std::move(key), std::move(edge.get_props()));
+    if (indexes_.empty()) {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
+            auto partId = partEdges.first;
+            std::vector<kvstore::KV> data;
+            std::for_each(partEdges.second.begin(), partEdges.second.end(), [&](auto& edge){
+                VLOG(4) << "PartitionID: " << partId << ", VertexID: " << edge.key.src
+                        << ", EdgeType: " << edge.key.edge_type << ", EdgeRanking: "
+                        << edge.key.ranking << ", VertexID: "
+                        << edge.key.dst << ", EdgeVersion: " << version;
+                auto key = NebulaKeyUtils::edgeKey(partId, edge.key.src, edge.key.edge_type,
+                                                   edge.key.ranking, edge.key.dst, version);
+                data.emplace_back(std::move(key), std::move(edge.get_props()));
+            });
+            doPut(spaceId_, partId, std::move(data));
         });
-        doPut(spaceId, partId, std::move(data));
+    } else {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
+            auto partId = partEdges.first;
+            const auto &edges = partEdges.second;
+            auto atomic = [&]() -> std::string {
+                return addEdges(version, partId, edges);
+            };
+            auto callback = [partId, this](kvstore::ResultCode code) {
+                handleAsync(spaceId_, partId, code);
+            };
+            this->kvstore_->asyncAtomicOp(spaceId_, partId, atomic, callback);
+        });
+    }
+}
+
+std::string AddEdgesProcessor::addEdges(int64_t version, PartitionID partId,
+                                        const std::vector<cpp2::Edge>& edges) {
+    std::unique_ptr<kvstore::BatchHolder> batchHolder = std::make_unique<kvstore::BatchHolder>();
+    std::map<std::string, std::pair<std::string,
+             std::pair<cpp2::EdgeKey, cpp2::IndexItem>>> newIndexes;
+    std::for_each(edges.begin(), edges.end(), [&](auto& edge){
+        auto prop = edge.get_props();
+        auto type = edge.key.edge_type;
+        auto srcId = edge.key.src;
+        auto rank = edge.key.ranking;
+        auto dstId = edge.key.dst;
+        VLOG(4) << "PartitionID: " << partId << ", VertexID: " << srcId
+                << ", EdgeType: " << type << ", EdgeRanking: " << rank
+                << ", VertexID: " << dstId << ", EdgeVersion: " << version;
+        auto key = NebulaKeyUtils::edgeKey(partId, srcId, type, rank, dstId, version);
```

Comment:
When inserting an edge,   the operations looks like:
1.  Get the original value for the edgeKey.
2.  Delete the old index for the related indexId
3.  Insert the new index and data.